transformer_engine/pytorch/attention.py

```diff
@@ -436,6 +436,9 @@ def get_attention_backend(
                 "padding between sequences, i.e. [a, a, PAD, b, b, b, PAD, c, PAD]"
             )
             use_flash_attention = False
+        if use_flash_attention and _use_flash_attn_3 and fp8 and fp8_meta["recipe"].fp8_dpa:
+            logger.debug("Disabling FlashAttention 3 for FP8 and qkv_format = thd")
+            _use_flash_attn_3 = False
```

Comment:
qkv_format == sbhd/bshd and "padding" in attn_mask_type also cannot run FP8, right?
and at here, I think you can directly set `use_flash_attention = False`, because FA2 cannot run FP8 anyway.

Reply:
Maybe just add `if use_flash_attention and _use_flash_attn_3 and (qkv_format == "thd" or "padding" in attn_mask_type): use_flash_attention = False` after [this line](https://github.com/cyanguwa/TransformerEngine/blob/add_descales/transformer_engine/pytorch/attention.py#L395).

Reply:
Thanks for helping me figure out the logic here :) It's a bit of a brain teaser I guess.

I almost went with what you said, but on second thought, I feel my current logic is right. Hear me out: 
- For F16 and the `qkv_format == sbhd/bshd and "padding" in attn_mask_type` case, we allow the `FlashAttention` backend (i.e. `use_flash_attention=True`), but not the `flash_attn_func` call from FA2. Because `FlashAttention` can pack up the useful tokens through `pack_tensors()` and then call `flash_attn_varlen_func`.
- For FP8, we can't even do that because `flash_attn_varlen_func` doesn't support FP8 yet.

Let me know if this makes sense. Thanks.

Reply:
Sorry, I do not totally understand this.

You are checking under `if qkv_format == "thd"`,  and you have `and fp8 and fp8_meta["recipe"].fp8_dpa`, which means only FP8 with THD format running can come to the code at here, right? neither FA2 nor FA3 can support FP8 with THD format, right? so I think we can directly set `use_flash_attention=False` at here (not just disable FA3)?

`qkv_format == sbhd/bshd and "padding" in attn_mask_type` will skip the code at here because `qkv_format != thd`. I think we mainly want to check if the FP8 running case can be supported by FlashAttention, so I [made this recommendation](https://github.com/NVIDIA/TransformerEngine/pull/1174/files#r1786920097). It's a check added under `if fp8 and fp8_meta["recipe"].fp8_dpa`, so should not impact BF16 at all, right?

Reply:
We disable FP8 for FA2 [here](https://github.com/cyanguwa/TransformerEngine/blob/b765f3d5c042b7adc79fdff4a29a66fbd20c68c1/transformer_engine/pytorch/attention.py#L388) under `Filter: Execution type`, and it'll never support FP8.

We disable FP8 THD for FA3 [here](https://github.com/cyanguwa/TransformerEngine/blob/b765f3d5c042b7adc79fdff4a29a66fbd20c68c1/transformer_engine/pytorch/attention.py#L439) under `Filter: QKV layout`, but FA3 might add the support soon, given how fast things are moving.

I'd like to keep things separate under different `Filter`s and also give FA3 the room to change (without us changing too much).

With `Filter: attn_mask_type`, we do support `qkv_format == sbhd/bshd and "padding" in attn_mask_type` with the `FlashAttention` path, hence `use_flash_attention` can be True. But within it, we have to distinguish the functionality of `flash_attn_func` and `flash_attn_varlen_func`, and their difference in FA2 and FA3 as well.

Reply:
I expect to remove [this](https://github.com/cyanguwa/TransformerEngine/blob/b765f3d5c042b7adc79fdff4a29a66fbd20c68c1/transformer_engine/pytorch/attention.py#L439) section, and the `or fp8` [here](https://github.com/cyanguwa/TransformerEngine/blob/b765f3d5c042b7adc79fdff4a29a66fbd20c68c1/transformer_engine/pytorch/attention.py#L5014), when `flash_attn_varlen_func` in FA3 supports FP8. Nothing else should need to be changed (none of the mask related code).

Reply:
Resolved offline.